feat#22/로그인 로그아웃 로직 구현

kkyu0718 opened the pull request:
## 📄 작업 설명
- spring-session 적용하여 서블릿 컨테이너 세션이 아닌 애플리케이션 레벨에서의 세션을 사용 -> 구현체(redis, jdbc) 를 바꿀 때 용이함
- 로그인 로그아웃 로직 구현 및 테스트 코드 작성
- Dto 검증 테스트 추가
- AuthInterceptor 를 통한 인증 필터링 로직 구현
- AuthArguementResolver 통해 컨트롤러에서 Authentication 객체 인자로 받을 수 있도록 구현
- 서비스 테스트 코드 DCI 패턴으로 리팩토링

## 🚨 관련 이슈
closes #22 

## 🌈 작업 상황
- 로그인과 관련하여 비밀번호 대조 로직이 없음 -> 추후에 OAuth 대체 가능성 생각하여 구현을 늦춤

kkyu0718 committed:
feat: 유저 로그인 서비스 로직 구현

email 를 받아 존재하는 유저인지만 검증하고 인증으로 처리 (비밀번호 대조 x)

AuthenticatlUtils 의 setMemberAuthenticated 를 구현하여 사용

kkyu0718 committed:
feat: spring-session 설정

서블릿 컨테이너 세션이 아닌 어플리케이션 레벨에서의 세션을 사용

추후에 구현체를 바꾸는데 용이함

kkyu0718 committed:
feat: Session 유틸 클래스 구현

session 과 authentiation 에 직접 접근 가능한 유틸 메소드 제공

kkyu0718 committed:
feat: auth 에러코드 추가

로그인 실패 에러코드가 포함된 auth 도메인 에러코드 구현

kkyu0718 committed:
feat: 유저 로그인 로직 구현

로그인에 성공시 SessionUtils 를 사용하여 authentication 클래스 저장

kkyu0718 committed:
feat: 로그인 서비스 로직 테스트 코드 추가

kkyu0718 committed:
feat: 로그아웃 서비스 로직 구현

로그아웃 시 SessionUtils 의 invalidateSession 으로 HttpSession 를 무효화시킴

kkyu0718 committed:
feat: 로그아웃 서비스 로직 테스트 코드 추가

kkyu0718 committed:
feat: Auth 컨트롤러 추가

kkyu0718 committed:
feat: Auth 컨트롤러 테스트 작성

kkyu0718 committed:
docs: 로그인 로그아웃 api 문서화

kkyu0718 committed:
refactor: AuthController HttpStatus 응답값 명시

@ResponseStatus 를 통해 디폴트가 OK 여도 명시하기로 합의되어 수정

kkyu0718 committed:
refactor: LoginRequestDto 클래스명 변경

dto 패키지 안에 있기 때문에 Dto 를 클래스명에서 제외

kkyu0718 committed:
feat: LoginRequest 필드 검증 로직 추가

kkyu0718 committed:
refactor: dto 패키지 auth 도메인 안으로 이동

kkyu0718 committed:
feat: LoginRequest 유효성 검사 테스트 추가

ValidationFactory 사용하여 유효한 요청 확인

kkyu0718 committed:
refactor: AuthServiceTest DCI 패턴으로 변경

kkyu0718 committed:
refactor: MemberServiceTest DCI 패턴으로 변경

kkyu0718 committed:
fix: 로그인 실패시 에러코드 401 -> 400 수정

로그인 실패는 클라이언트의 잘못된 요청이기 때문에 400 으로 수정

kkyu0718 committed:
refactor: SessionUtils 의 메소드 AuthenticationUtils 클래스로 이동

SessionUtils 안에서 authentication 관련된 메소드는 AuthenticationUtils 로 이동

kkyu0718 committed:
feat: AuthInterceptorTest 구현

Authentication 유효성 검증 테스트

kkyu0718 committed:
feat: AuthInteceptor 인증 인터셉터 구현

authentication 객체에 저장된 유저 정보를 db 와 대조하는 로직 구현

kkyu0718 committed:
refactor: AuthenticationContext 삭제 후 AuthenticationUtils 로 수정

AuthenticationUtils 를 쓰게됨에 따라 AuthenticationContext 를 삭제하고 수정

kkyu0718 committed:
feat: AuthenticationUtils 세션 비활성화 코드 추가

kkyu0718 committed:
refactor: 테스트 코드 실행 전 테스트 유저 저장하도록 수정

given 를 nested 로 밖으로 빼 공통적인 로직을 응집

kkyu0718 committed:
style: AuthenticationInterceptor 로깅 추가

kkyu0718 committed:
fix: AuthInterceptorTest 잘못된 테스트코드 수정

- Authentication 인자 순서 수정
- 잘못된 회원 정보와 비교할 수 있도록 인자 수정

kkyu0718 committed:
feat: AuthArgumentResolver 로직 구현

로그인을 한 유저가 컨트롤러에서 Authentication 인자로 인증객체에 접근할 수 있도록 구현

kkyu0718 committed:
feat: WebMvcConfig 로직 구현

AuthInterceptor, AuthArgumentResolver 등록

kkyu0718 committed:
style: LoginRequestTest 한글 테스트명 추가

kkyu0718 committed:
fix: 유저 회원가입 api url 변경

회원가입은 인증이 필요없고 회원탈퇴는 인증이 필요함

원래의 회원가입 POST /member 가 회원탈퇴 DELETE /member 와 겹쳐서 interceptor 제외 url 로 분류해낼 수 없는 이슈 발생하여 수정

kkyu0718 committed:
fix: RestDocsSupport @MockBean MemberRepository 추가

WebMvcConfig 빈 생성 에러가 발생하여 의존성 모킹 필요

kkyu0718 committed:
Merge branch 'develop' into feat#22/auth

HyeonJun0530 commented on src/main/java/com/wootecam/festivals/domain/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,34 @@
+
+    @ResponseStatus(HttpStatus.OK)
+    @PostMapping("/login")
+    public ApiResponse<Void> login(@RequestBody LoginRequest dto) {
```
@Valid 어노테이션이 빠져있네요 !

HyeonJun0530 commented on src/main/java/com/wootecam/festivals/domain/auth/dto/LoginRequest.java:
```diff
@@ -0,0 +1,10 @@
+import jakarta.validation.constraints.Size;
+
+public record LoginRequest(
+        @NotBlank(message = "이메일은 필수입니다.")
+        @Size(min = 1, max = 100, message = "이메일은 1자 이상 100자 이하여야 합니다.")
+        String email) {
```
@Email 이라는 어노테이션을 사용해 Email을 Valid 하는 것이 좋아보이네요 !
혹시 Size를 사용한 다른 이유가 있으실까요?

kkyu0718 replied:
email 어노테이션 추가하도록 하겠습니다! 저희가 별도의 유효성 검증 없이 (실제 이메일인지) 이메일을 받고 있어서 비정상적으로 길이가 긴 데이터에 대한 검증 정도는 필요하다고 생각했어요!

-----추가된 내용------
이메일 어노테이션 확인해보니 @ 이전까지에 대한 길이는 64자 이하로 제한하지만 @ 이후에 대해서는 길이 검증을 안하네요!!

HyeonJun0530 commented on src/main/java/com/wootecam/festivals/global/auth/AuthArgumentResolver.java:
```diff
@@ -0,0 +1,28 @@
+
+public class AuthArgumentResolver implements HandlerMethodArgumentResolver {
+
+    @Override
+    public boolean supportsParameter(MethodParameter parameter) {
+        return parameter.getParameterType().equals(Authentication.class);
+    }
+
```
어노테이션 방식말고 이렇게 클래스 방식을 쓰셨는데 어떤 점 때문에 쓰신 건가요?

kkyu0718 replied:
말씀하신 방법이 더 좋아보여서 수정했습니다!

HyeonJun0530 commented on src/main/java/com/wootecam/festivals/global/config/WebMvcConfig.java:
```diff
@@ -0,0 +1,35 @@
+    @Override
+    public void addInterceptors(InterceptorRegistry registry) {
+        registry.addInterceptor(new AuthInterceptor(memberRepository))
+                .addPathPatterns("/**")
+                .excludePathPatterns("**/public/**"
+                        ,"/api/*/auth/login"
+                        , "/api/*/member/signup"
+                        , "**/error");
+    }
```
Interceptor를 bean으로 등록하지 않고 new 통해 생성해주는 이유가 무엇일까요?
bean으로 하면 MemberRepository도 필드로 존재하지 않아도 될텐데 특별한 이유가 있으신건가요?

pathPatterns 부분은 현재는 모든 로그인이 성공하는 상태라 나중에 바뀔 땐 이 부분도 상의해봐야겠네요 !

kkyu0718 replied:
빈 등록 방식으로 변경하였습니다. 덕분에 memberRepository 관련 mock 다 뺐어요!

HyeonJun0530 commented on src/test/java/com/wootecam/festivals/docs/utils/RestDocsSupport.java:
```diff
@@ -25,6 +27,9 @@
 @ExtendWith(RestDocumentationExtension.class)
 public abstract class RestDocsSupport {
 
+    @MockBean
+    public MemberRepository memberRepository;
```
이 MockBean은 무슨 역할을 하는지 알 수 있을까요?

kkyu0718 replied:
WebMvcTest 가 자동으로 WebMvcConfig 와 AuthInterceptor 를 자동으로 불러오는데 AuthInterceptor 가 MemberRepository 를 의존하기 때문에 모킹으로 해결했었어요! 

이번에 수정하면서 WebMvcTest 가 자동으로 WebMvcConfig 와 AuthInterceptor 를 불러오지 않도록 수정했습니다.

HyeonJun0530 commented on src/test/java/com/wootecam/festivals/domain/auth/controller/AuthControllerTest.java:
```diff
@@ -0,0 +1,91 @@
+                .content(objectMapper.writeValueAsString(loginRequest))
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andDo(restDocs.document(
+                        requestFields(
+                                fieldWithPath("email").description("로그인할 이메일")
+                        ),
```
제약 조건도 attribute나 optional 같은 걸로 명시해서 문서에 한번에 볼 수 있게끔 하면 좋겠어요 !

HyeonJun0530 commented on src/test/java/com/wootecam/festivals/domain/auth/dto/LoginRequestTest.java:
@ParameterizeTest를 활용하면 좋을 것 같아요 !
현종님이 작성한 Festival 테스트 쪽을 보시면 감이 잡히실꺼예요 !

kkyu0718 replied:
이거 반영해서 수정했습니다 감사해요~!

HyeonJun0530 reviewed (changes requested):
리뷰 작성한 코드를 제외하고는 나머지 코드는 큰 문제 없어 보이네요 👍 
리뷰도 그렇게 한 이유가 있다면 괜찮을 거 같아요 !
천천히 보시고 답글 부탁드려요 😊

kkyu0718 committed:
feat: LoginRequest @Email 검증 추가 및 테스트

email 검증 어노테이션을 추가하여 검증 고도화

kkyu0718 committed:
refactor: LoginRequestTest DCI 패턴으로 변경

kkyu0718 committed:
refactor: AuthInterceptor 빈 등록

WebMvcConfig 에 자동으로 의존성 주입하기 위해 빈 등록

kkyu0718 committed:
fix: WebMvcTest 의 WebMvcConfig 의존성 주입 관련 에러 해결

WebMvcTest 가 자동으로 WebMvcConfig 와 Interceptor 를 불러오기 때문에 두 개의 클래스가 필요한 의존성이 스캔안되어 에러 발생

프로필이 test 일 경우에는 동작하지 않도록 어노테이션 설정

kkyu0718 committed:
refactor: AuthArgumentResolver 어노테이션 방식으로 수정

AuthUser 라는 어노테이션으로 컨트롤러에서 접근 가능하도록 수정

kkyu0718 committed:
docs: 로그인 로직 제약 사항 추가

kkyu0718 committed:
Merge branch 'develop' into feat#22/auth

kkyu0718 commented:
WebMvcConfig 와 AuthInterceptor 테스트 환경에서는 등록되지 않도록 해뒀으니 참고해주세요~ 
@HyeonJun0530 @minnim1010 @bellringstar

kkyu0718 committed:
fix: TicketController ActiveProfile 추가

profile test 로 인식되지 못하는 문제점 때문에 직접 추가

kkyu0718 committed:
fix: auth 유저 추가 후 인증 필요한 코드 수정

인증 객체를 컨트롤러에서 받을 수 있게 되어 기존의 코드 수정
